### PR TITLE
Add Arc shape control 2: Electric Boogaloo

### PR DESCRIPTION
**Pull request by Splitwirez**
## What does the pull request do?
Adds a new `Arc` control to `Avalonia.Controls.Shapes`, which is needed for the `ProgressRing` control which I am also working on ( https://github.com/Splitwirez/Avalonia/tree/progress-ring , [NEW PR TBD] ), and may also see some use elsewhere.

### CODE REVIEW

File: src/Avalonia.Controls/Shapes/Arc.cs
Comment (jmacato) [inline src/Avalonia.Controls/Shapes/Arc.cs:69]: Use `var`

File: src/Avalonia.Controls/Shapes/Arc.cs
Comment (jmacato) [inline src/Avalonia.Controls/Shapes/Arc.cs:65]: Format this comment

File: src/Avalonia.Controls/Shapes/Arc.cs
Comment (jmacato) [inline src/Avalonia.Controls/Shapes/Arc.cs:46]: `var`

File: src/Avalonia.Controls/Shapes/Arc.cs
Comment (jmacato) [inline src/Avalonia.Controls/Shapes/Arc.cs:40]: use `=>` instead of the full fat getter/setter

**Review by jmacato — CHANGES_REQUESTED** [review]
almost there :) needs examples in RenderDemo too

**Comment by robloo** [comment]
Same comments as here: https://github.com/AvaloniaUI/Avalonia/pull/6003#issuecomment-856046022. Not trying to throw a wrench in things but the API should be looked at carefully for brand new code.

> I also looked in more detail at the Arc shape. Most arcs don't specify both a start and stop angle. Instead they specify the start and then a +/- sweep angle. I think it would be a good idea to change this for convention.

 * https://docs.microsoft.com/en-us/dotnet/api/system.windows.media.arcsegment?view=net-5.0
 * https://docs.microsoft.com/en-us/xamarin/xamarin-forms/user-interface/graphics/skiasharp/curves/arcs
 * https://developer.android.com/reference/android/graphics/drawable/shapes/ArcShape
 * https://docs.microsoft.com/en-us/uwp/api/windows.ui.xaml.media.arcsegment?view=winrt-20348

**Comment by maxkatz6** [comment]
@robloo I see these Start and Sweep properties in this PR

**Comment by robloo** [comment]
Well, I'm dumb. I should have actually checked in detail and not assumed it was carry over.

File: src/Avalonia.Controls/Shapes/Arc.cs
Comment (robloo) [inline src/Avalonia.Controls/Shapes/Arc.cs:38]: Old summary comments?

File: src/Avalonia.Controls/Shapes/Arc.cs
Comment (robloo) [reply src/Avalonia.Controls/Shapes/Arc.cs:69]: var was originally added to make Linq usage more sane. For basic types line this it can actually make code less readable.

File: src/Avalonia.Controls/Shapes/Arc.cs
Comment (Splitwirez) [reply src/Avalonia.Controls/Shapes/Arc.cs:40]: ...wait, seriously? I only used `get { return` instead of `get =>` because that's what the rest of Avalonia itself seems to do everywhere I look. I'll make the change, but...am I missing something here?

File: src/Avalonia.Controls/Shapes/Arc.cs
Comment (Splitwirez) [reply src/Avalonia.Controls/Shapes/Arc.cs:38]: I *think* I updated that one...though given you had to ask, I'm guessing I didn't do a very good job of it. So uhhh...how would you explain the sweep angle property?

**Comment by Splitwirez** [comment]
> Same comments as here: [#6003 (comment)](https://github.com/AvaloniaUI/Avalonia/pull/6003#issuecomment-856046022). Not trying to throw a wrench in things but the API should be looked at carefully for brand new code.
> 
> > I also looked in more detail at the Arc shape. Most arcs don't specify both a start and stop angle. Instead they specify the start and then a +/- sweep angle. I think it would be a good idea to change this for convention.
> 
>     * https://docs.microsoft.com/en-us/dotnet/api/system.windows.media.arcsegment?view=net-5.0
> 
>     * https://docs.microsoft.com/en-us/xamarin/xamarin-forms/user-interface/graphics/skiasharp/curves/arcs
> 
>     * https://developer.android.com/reference/android/graphics/drawable/shapes/ArcShape
> 
>     * https://docs.microsoft.com/en-us/uwp/api/windows.ui.xaml.media.arcsegment?view=winrt-20348

Understandable oversight tbh, lol. But yes, I did make the changes while I was setting up the new branch.

File: src/Avalonia.Controls/Shapes/Arc.cs
Comment (Gillibald) [reply src/Avalonia.Controls/Shapes/Arc.cs:69]: I think this is more about code style. If the code base usually uses `var` that should be followed.

File: src/Avalonia.Controls/Shapes/Arc.cs
Comment (Gillibald) [reply src/Avalonia.Controls/Shapes/Arc.cs:40]: I think all new definitions use the arrow function style. We can't touch old definition just to change the code style.

File: src/Avalonia.Controls/Shapes/Arc.cs
Comment (robloo) [reply src/Avalonia.Controls/Shapes/Arc.cs:69]: It is entirely possible to overuse var. In this case it isn't obvious if the type is float or double unless you have a lot of experience with the objects. This is especially true for those of us that use WinUI where Microsoft is changing to float. There is no reason to hide the double type in this case, it isnt too verbose.

File: src/Avalonia.Controls/Shapes/Arc.cs
Comment (robloo) [reply src/Avalonia.Controls/Shapes/Arc.cs:40]: Yea, in this case I do see a lot of code moving to `=>`. This was done in my code base as well and is one of those things that just simplifies things.

> We can't touch old definition just to change the code style.

Um, what!? Refactoring just to refactor, clean-up and improve code is very important with software.

File: src/Avalonia.Controls/Shapes/Arc.cs
Comment (robloo) [reply src/Avalonia.Controls/Shapes/Arc.cs:38]: On second look I see how it could be correct, still a little confusing for me.

> how would you explain the sweep angle property?

What about: "Gets or sets the angle, in degrees, added to the <see cref="StartAngle"/> defining where the arc ends. A positive value is clockwise, negative is counter-clockwise."

File: src/Avalonia.Controls/Shapes/Arc.cs
Comment (grokys) [reply src/Avalonia.Controls/Shapes/Arc.cs:40]: > Um, what!? Refactoring just to refactor, clean-up and improve code is very important with software.

Not when you have to do backports/git blame often as we do. See https://brennan.io/2021/06/15/git-less-is-more/ for a discussion.

Comment (robloo) [reply src/Avalonia.Controls/Shapes/Arc.cs:40]: This is very concerning for the long term health of the project. Actively discouraging refactoring should not be done.

File: src/Avalonia.Controls/Shapes/Arc.cs
Comment (robloo) [reply src/Avalonia.Controls/Shapes/Arc.cs:69]: I dont know how you can disagree with this. Almost any discussion, article or commentary on the use of var will back up my points.

File: src/Avalonia.Controls/Shapes/Arc.cs
Comment (grokys) [reply src/Avalonia.Controls/Shapes/Arc.cs:40]: We're not actively discouraging refactoring, there are just 2 things to consider:

1. Who wants to do it
2. When to do it

So far, no-one has come forward for 1) so we've not discussed 2).

And having mixed syntax for getters and setters is inconsistent, sure, but I don't see how it's "concerning for the health of the project". There are much more important refactors to be done, such as Nullable Reference Types.

File: src/Avalonia.Controls/Shapes/Arc.cs
Comment (grokys) [reply src/Avalonia.Controls/Shapes/Arc.cs:69]: _I_ certainly can disagree with it, and when I've seen it discussed on twitter etc, it appears to me to entirely come down to taste. Personally I much prefer `var`.

Comment (robloo) [reply src/Avalonia.Controls/Shapes/Arc.cs:69]: If this direction is set by you, I can't change it. You founded the code and can firmly set the standard for things like this. However, var is not intended for use everywhere like in a non-typesafe scripting language. It would be wise to let developers use it as they see best and error on the side of not-using var to ensure types are readable and clear. I've given examples for this case (float vs double). The arguments of my viewpoint are objectively stronger than simply requiring var for consistency and code style.

File: src/Avalonia.Controls/Shapes/Arc.cs
Comment (grokys) [reply src/Avalonia.Controls/Shapes/Arc.cs:40]: I can only speak for myself really: because none of the core contributors have brought this up recently we've not got an official line on this.

To give some background: at the start of the project I used StyleCop because I wanted to enforce a strict coding style. This received at _lot_ of pushback from various contributors so I relented and we removed it. As the project grew, I became less concerned about uniform code style and more concerned about getting useful info quickly from `git blame` and being able to backport changes.

File: src/Avalonia.Controls/Shapes/Arc.cs
Comment (grokys) [reply src/Avalonia.Controls/Shapes/Arc.cs:69]: I agree, I wouldn't have flagged usage of non-`var` in this PR personally.

File: src/Avalonia.Controls/Shapes/Arc.cs
Comment (Splitwirez) [reply src/Avalonia.Controls/Shapes/Arc.cs:65]: ...er...what exactly do you mean by that?

File: src/Avalonia.Controls/Shapes/Arc.cs
Comment (robloo) [reply src/Avalonia.Controls/Shapes/Arc.cs:40]: That's fair. I'll shut up about this topic 😄

Interesting background as well. Personally, I keep code analyzers enabled but disable rules that are too unforgiving. There are inevitably special cases as well. The biggest issue now is build performance as analyzers significantly increase build time.

Comment (Splitwirez) [reply src/Avalonia.Controls/Shapes/Arc.cs:40]: ...to clarify, I didn't intend to *challenge* this change request, I was just surprised by it, and had wondered if there may have been some mistake, so I asked for clarification before fulfilling it.

On my own projects, I code by my own rules. Avalonia, however, is not one of my own projects, and so I will gladly play by the rules of the project's core maintainers or whoever else is in charge. _If_ the request pertained to a matter I was more knowledgeable about _and_ I saw a potential drawback of the request, I wouldn't hesitate to bring it up, but the significance of code style is not a matter I am particularly knowledgeable on, so I don't expect that to happen here. Either way though, I respect that ultimately, it's not my call to make.

Comment (grokys) [reply src/Avalonia.Controls/Shapes/Arc.cs:40]: I just wish C# had something like `go fmt` so we could just avoid these discussions ;)

File: src/Avalonia.Controls/Shapes/Arc.cs
Comment (Splitwirez) [reply src/Avalonia.Controls/Shapes/Arc.cs:69]: ...so is there a final decision on this yet?

File: src/Avalonia.Controls/Shapes/Arc.cs
Comment (Splitwirez) [reply src/Avalonia.Controls/Shapes/Arc.cs:38]: That sounds much better to me, yeah.

File: src/Avalonia.Controls/Shapes/Arc.cs
Comment (workgroupengineering) [reply src/Avalonia.Controls/Shapes/Arc.cs:40]: @grokys exits equivalent of go fmt and it is https://github.com/dotnet/format

File: src/Avalonia.Controls/Shapes/Arc.cs
Comment (FoggyFinder) [reply src/Avalonia.Controls/Shapes/Arc.cs:65]: I suppose it should be with a leading whitespace

```cs
else // partial arc
```

File: src/Avalonia.Controls/Shapes/Arc.cs
Comment (jmacato) [reply src/Avalonia.Controls/Shapes/Arc.cs:69]: @Splitwirez use var as i suggested. That's how our codebase is.

Comment (robloo) [reply src/Avalonia.Controls/Shapes/Arc.cs:69]: I strongly recommend you change this requirement as it reduces readability in cases like this and hides information. I can even prove it violates the design intent of C# and Microsofts guidelines.

 If I commit code to the repo expect another conversation about it. Otherwise, 

 * https://www.researchgate.net/profile/Pierre-Akiki/publication/328999651_To_var_or_not_to_var_how_do_C_developers_use_and_misuse_implicit_and_explicit_typing/links/5c3b160c299bf12be3c4334f/To-var-or-not-to-var-how-do-C-developers-use-and-misuse-implicit-and-explicit-typing.pdf
 * https://intellitect.com/when-to-use-and-not-use-var-in-c/

These arguments are immensely stronger than those for consistency.

File: src/Avalonia.Controls/Shapes/Arc.cs
Comment (robloo) [inline src/Avalonia.Controls/Shapes/Arc.cs:94]: `0 +` why add zero?

File: src/Avalonia.Controls/Shapes/Arc.cs
Comment (maxkatz6) [reply src/Avalonia.Controls/Shapes/Arc.cs:69]: Let's close this conversation about style guides in this PR. 
In general I prefer consistency with whatever approach was actually used and whatever advantages or disadvantages it has. Still code style question does not affects neither Avalonia behavior nor user applications behavior.
So, if we want to discuss code style questions and enforce some specific rules (even with https://github.com/dotnet/format) we should continue in separated discussion.

And as for this PR, we should ignore this problem and unblock contributors.

File: src/Avalonia.Controls/Shapes/Arc.cs
Comment (maxkatz6) [reply src/Avalonia.Controls/Shapes/Arc.cs:46]: Marked as resolved, see https://github.com/AvaloniaUI/Avalonia/pull/6133#discussion_r660799364

File: src/Avalonia.Controls/Shapes/Arc.cs
Comment (maxkatz6) [reply src/Avalonia.Controls/Shapes/Arc.cs:40]: Marked as resolved, see https://github.com/AvaloniaUI/Avalonia/pull/6133#discussion_r660799364

File: src/Avalonia.Controls/Shapes/Arc.cs
Comment (Splitwirez) [reply src/Avalonia.Controls/Shapes/Arc.cs:84]: > Without testing this code, I think you need to set the SweepDirection to CounterClockwise if the sweep angle is negative. Hardcoding this to Clockwise will likely only work for positive values.

I already flip them around earlier in `CreateDefiningGeometry`, which solves that problem.

> With that said, this whole class might need to be tested with negative startt and sweep angles. The start angle should be properly normalized so -90 becomes 270 (I think that works) and the sweep angle should be using the normalized absolute value but the sweep direction must be set based on the negative sign.

They're both already normalized in `CreateDefiningGeometry`. Is that...meaningfully different from what you're suggesting here?

**Comment by Splitwirez** [comment]
> almost there :) needs examples in RenderDemo too

@jmacato Okay...what kind of example do you have in mind? I have an `ArcPage` on hand, which I'd been using to test in the control catalog (didn't know about RenderDemo)...can I just label the sliders on that and toss it into RenderDemo, or do you have something more specific in mind?

File: src/Avalonia.Controls/Shapes/Arc.cs
Comment (robloo) [reply src/Avalonia.Controls/Shapes/Arc.cs:84]: Ah, ok, I should have looked in more detail above. I just saw the hard-coded clockwise sweep direction. I have no concerns with this now.

**Comment by jmacato** [comment]
@Splitwirez ArcPage seems good 👍

**Comment by Takoooooo** [comment]
This may be helpful for you.
https://github.com/PieroCastillo/Aura.UI/blob/master/src/Aura.UI/Controls/Colouring/Arc/Arc.cs
Code looks much dirtier but this alternative implementation may be better in terms of performance

**Comment by grokys** [comment]
Are we happy with this now? The discussion on this has been rather excessive for such a simple PR :)

**Review by jmacato — APPROVED** [review]
The samples can be done on another PR. LGTM!

**Pull request merged**
